### PR TITLE
Use System Encoding for java command line argument files (fixes #45)

### LINKED ISSUE #45: [win] can not launch java with long classpath containing non-Ascii letter
Given a workspace in a path with a non-Ascii letter like "C:\Jörg"
And then try to launch a plain test.Main.main() the command line will be like:
`C:\Program Files\Eclipse Adoptium\jdk-17.0.3.7-hotspot\bin\java.exe -classpath C:\Jörg\projectName\test\bin test.Main`  (OK)
but when the classpath gets too long for command line then the class path argument is wrapped in a UTF-8 encoded file like
"argument.txt":
`-classpath C:\Jörg\projectName\test\bin`  (OK)
and the command line will be
`C:\Program Files\Eclipse Adoptium\jdk-17.0.3.7-hotspot\bin\java.exe "@argument.txt" test.Main`  (OK)
that will result in "java.lang.ClassNotFoundException: test.Main"  (NOK)
If i manually convert the argument.txt from UTF8 to ANSI then the command line works. 
The argument file has always to be encoded with the system encoding instead of UTF-8.
see
https://docs.oracle.com/en/java/javase/17/docs/specs/man/java.html#java-command-line-argument-files
"_The argument file must contain only ASCII characters or characters in **system default encoding** that's ASCII friendly, such as UTF-8._"

Side note: in the launch configuration dialog is a button to show command line. But the copied command line does not directly work: 1. it is missing double quotes around the executable path ("Eclipse Adoptium" contains a white space) 
2. there are line breaks instead of blanks as delimiters.

The error is probably "StandardCharsets.UTF_8" in org.eclipse.jdt.internal.launching.ClasspathShortener.createClassPathArgumentFile(String) and org.eclipse.jdt.internal.launching.ClasspathShortener.createModulePathArgumentFile(String).

from Bug 385738 - Shorten the classpath/modulepath if necessary.

### PR DESCRIPTION
**Pull request by jukzi**
Otherwise encoding is wrong and application may not start.

see https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/45

### CODE REVIEW

**Comment by vogella** [comment]
@jjohnstn can you review?

File: org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/ClasspathShortener.java [hunk 1]
```diff
@@ -55,6 +55,11 @@
  *
  */
 public class ClasspathShortener {
+	// We need system encoding see https://docs.oracle.com/en/java/javase/18/docs/specs/man/java.html#java-command-line-argument-files
+	// Implementation from https://openjdk.java.net/jeps/400
+	static final Charset SYTEM_CHARSET = (System.getProperty("native.encoding") != null) // Populated on Java 18 and later //$NON-NLS-1$
```
Comment (jjohnstn) [inline org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/ClasspathShortener.java:60]: typo: Should be SYSTEM_CHARSET

**Review by jjohnstn — CHANGES_REQUESTED** [review]

**Review by jjohnstn — APPROVED** [review]

**Comment by jjohnstn** [comment]
@iloveeclipse I think this patch is fine and uses the suggested logic in the JEP documentation.  Since you were commenting on the original issue, I will ask if you have any other comments, otherwise, it can be merged.

**Comment by iloveeclipse** [comment]
I'm not sure I can see any change with the PR. 

I've patched `org.eclipse.jdt.internal.launching.ClasspathShortener.getMaxCommandLineLength()` to return 42 on Windows  and see that the file is still generated with UTF-8 on Java 11 in `org.eclipse.jdt.internal.launching.ClasspathShortener.createClassPathArgumentFile(String)`.

Using Java 11 I can't launch the class, using Java 8 everything works.

Here is what I get from java 11 (java 17 similar)
```
Error: Could not find or load main class @D:\tmp\wsp411\JÃ¶rg\.temp-Joerg-classpath-arg-1652700414353.txt
Caused by: java.lang.ClassNotFoundException: @D:\tmp\wsp411\JÃ¶rg\.temp-Joerg-classpath-arg-1652700414353.txt
```

I'm missing something? Ideally we would have a test that shows the expected vs current behavior.

**Review by iloveeclipse — CHANGES_REQUESTED** [review]
See my comment on PR. I assume something is still missing, or at least a test that shows the change actually works as expected (might be I'm testing it wrong manually).

**Comment by iloveeclipse** [comment]
Here my project which I've used to see what is working what is not: [pr_47.zip](https://github.com/eclipse-jdt/eclipse.jdt.debug/files/8699643/pr_47.zip). I couldn't get it working on Java 11/17 on my (Windows 11) system in the workspace that reported UTF-8 as default encoding.

**Comment by jukzi** [comment]
if you start eclipse with `-Dfile.encoding=UTF-8`  that parameter is also passed to launched configurations.
Charset.defaultCharset() then NOT returns the SYSTEM encoding but the UTF-8 encoding passed. 
1. that's seems to be against the specification
2. why does eclipse start with that  -Dfile.encoding=UTF-8 by default?
The encoding can be changed in the launch configuration but unfortunatly not to SYSTEM default.
image

**Comment by iloveeclipse** [comment]
See https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/24

**Comment by jukzi** [comment]
jdk uses "sun.jnu.encoding" for real System charset
https://github.com/openjdk/jdk/pull/8378/files#diff-f9ae2535b3ccc7376de825e468d3e5758c0f476436d6f326361eb57d103f7e5aR155

**Comment by jukzi** [comment]
@iloveeclipse 
I changed to implementation that works even when -Dfile.encoding is set.
Also gives a reasonsanble error message if argument cannot be encoded:
image

please review

**Comment by jukzi** [comment]
ok, enjoy your vacation!!

File: org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/ClasspathShortener.java [hunk 2]
```diff
@@ -55,6 +55,15 @@
  *
  */
 public class ClasspathShortener {
+	// We need system encoding see https://docs.oracle.com/en/java/javase/18/docs/specs/man/java.html#java-command-line-argument-files
+	static final Charset SYTEM_CHARSET = (System.getProperty("native.encoding") != null) // Populated on Java 18 and later //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (iloveeclipse) [reply org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/ClasspathShortener.java:59]: [The PR with new API is merged](https://github.com/eclipse-platform/eclipse.platform/pull/38) and new API will be available after the next I-build.

**Comment by jukzi** [comment]
ping @iloveeclipse

**Comment by jukzi** [comment]
@iloveeclipse 
I think I covered all requested changes but still no review/submit, what do you think is missing?

**Review by iloveeclipse — APPROVED** [review]
LGTM

**Pull request merged**